### PR TITLE
Adding ECDH feature to ECC engine

### PR DESCRIPTION
**Pull request by mojtaba-bisheh**
Added ECDH feature to ECC block
Updated smoke tests to cover ECDH
fixed issue #556

### CODE REVIEW

File: src/integration/config/compile.yml
Comment (calebofearth) [inline src/integration/config/compile.yml:1]: This same rename is necessary in the [Makefile](https://github.com/chipsalliance/caliptra-rtl/blob/mojtabab-msft-ecdh/tools/scripts/Makefile#L85)

Comment (calebofearth) [reply src/integration/config/compile.yml:1]: And can you also update the README (which describes using this script)

Comment (mojtaba-bisheh) [reply src/integration/config/compile.yml:1]: Done!

**Review by calebofearth — APPROVED** [review]

File: src/ecc/tb/ecc_secp384r1.c [hunk 1]
```diff
@@ -439,15 +460,67 @@ int main( int argc, char *argv[] )
...
+    {
+        mbedtls_printf( "failed\n ! mbedtls_ecdh_compute_shared returned %d\n", ret );
+        // goto exit;
+    }
+    mbedtls_mpi_write_binary( &z, z_buf , z_buf_len );
+
+    // /*
+    //  * Compute ECDH shared key (Method 2 - reuse privkey)
+    //  */
+    // mbedtls_mpi_init( &z);
```
Comment (bluegate010) [inline src/ecc/tb/ecc_secp384r1.c:503]: Should this commented-out code be removed before merging?

File: src/integration/test_suites/smoke_test_kv_ecc_flow/smoke_test_kv_ecc_flow.c [hunk 2]
```diff
@@ -162,11 +162,30 @@ void main(){
...
+                                    0x1F802311,
+                                    0x21CCE3D3,
+                                    0x9BE52E00,
+                                    0xC194A413,
+                                    0x2C4A6C76,
+                                    0x8BCD94D2};
+
+    uint32_t ecc_pubkey_x_dh[] =   {0x793148F1,0X787634D5,0XDA4C6D90,0X74417D05,0XE057AB62,0XF82054D1,0X0EE6B040,0X3D627954,0X7E6A8EA9,0XD1FD7742,0X7D016FE2,0X7A8B8C66};
+    uint32_t ecc_pubkey_y_dh[] =   {0xC6C41294,0X331D23E6,0XF480F4FB,0X4CD40504,0XC947392E,0X94F4C3F0,0X6B8F398B,0XB29E4236,0X8F7A6859,0X23DE3B67,0XBACED214,0XA1A1D128};
+    uint32_t ecc_sharedkey_dh[] =  {0x5EA1FC4A,0XF7256D20,0X55981B11,0X0575E0A8,0XCAE53160,0X137D904C,0X59D926EB,0X1B8456E4,0X27AA8A45,0X40884C37,0XDE159A58,0X028ABC0E};
```
Comment (bluegate010) [inline src/integration/test_suites/smoke_test_kv_ecc_flow/smoke_test_kv_ecc_flow.c:181]: How were these test vectors generated?

**Review by bluegate010 — COMMENTED** [review]
Could this PR add some documentation on exactly what key derivation flow is being used to generate the final shared secret? I see the test code using an IV and nonce; how are those being used to derive the shared secret? Is there a pointer to some NIST document that this implementation is following?

**Comment by bluegate010** [comment]
> Could this PR add some documentation on exactly what key derivation flow is being used to generate the final shared secret? I see the test code using an IV and nonce; how are those being used to derive the shared secret? Is there a pointer to some NIST document that this implementation is following?

Specifically, I'm trying to gauge whether this implementation will gel with https://www.rfc-editor.org/rfc/rfc9629#section-5

File: src/integration/test_suites/smoke_test_kv_ecc_flow/smoke_test_kv_ecc_flow.c (hunk 2 above)
Comment (mojtaba-bisheh) [reply src/integration/test_suites/smoke_test_kv_ecc_flow/smoke_test_kv_ecc_flow.c:181]: they have been generated based on our mbedtls ref model

**Comment by mojtaba-bisheh** [comment]
> > Could this PR add some documentation on exactly what key derivation flow is being used to generate the final shared secret? I see the test code using an IV and nonce; how are those being used to derive the shared secret? Is there a pointer to some NIST document that this implementation is following?
> 
> Specifically, I'm trying to gauge whether this implementation will gel with https://www.rfc-editor.org/rfc/rfc9629#section-5

This PR only takes care of classic ECDH without KEM operation. any required changes for hybrid operation will be added later

**Comment by bluegate010** [comment]
> > > Could this PR add some documentation on exactly what key derivation flow is being used to generate the final shared secret? I see the test code using an IV and nonce; how are those being used to derive the shared secret? Is there a pointer to some NIST document that this implementation is following?
> > 
> > 
> > Specifically, I'm trying to gauge whether this implementation will gel with https://www.rfc-editor.org/rfc/rfc9629#section-5
> 
> This PR only takes care of classic ECDH without KEM operation. any required changes for hybrid operation will be added later

ECDH is a KEM. Hybrid is not required.

**Comment by mojtaba-bisheh** [comment]
> > > > Could this PR add some documentation on exactly what key derivation flow is being used to generate the final shared secret? I see the test code using an IV and nonce; how are those being used to derive the shared secret? Is there a pointer to some NIST document that this implementation is following?
> > > 
> > > 
> > > Specifically, I'm trying to gauge whether this implementation will gel with https://www.rfc-editor.org/rfc/rfc9629#section-5
> > 
> > 
> > This PR only takes care of classic ECDH without KEM operation. any required changes for hybrid operation will be added later
> 
> ECDH is a KEM. Hybrid is not required.

section 1 of RFC9629 describes KEM that needs ciphertext (ct) to be passed to the recipient. ECDH does not have encapsulate/decapsulate operations. 
this PR implements section 5.7.1.2 of SP800 56A https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf

**Comment by mojtaba-bisheh** [comment]
> > > ECDH is a KEM. Hybrid is not required.
> > 
> > 
> > section 1 of RFC9629 describes KEM that needs ciphertext (ct) to be passed to the recipient. ECDH does not have encapsulate/decapsulate operations. this PR implements section 5.7.1.2 of SP800 56A https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf
> 
> Thanks. I overreached when I said ECDH is a KEM - it's an ingredient, sure.
> 
> I perused section 5.7.1.2 of SP800 56A, and I see it describes an operation that computes an output based on domain parameters, a public key, and a private key. I don't see anything involving a nonce or IV. Yet, it appears from the tests in `src/integration/test_suites/libs/ecc/ecc.c` that a nonce and IV are part of the interface being added in this PR. Hence my questions as to what the implementation is doing with those arguments.

The IV is utilized as a source of randomness for the SCA countermeasure and does not alter the output. The inclusion of the nonce was an error due to a copy-paste from the key generation operation. I have removed the nonce from the smoke tests and will include a description of this mode in the specification.

**Review by Nitsirks — APPROVED** [review]

**Review by bluegate010 — APPROVED** [review]

**Review by upadhyayulakiran — APPROVED** [review]

**Pull request merged**